lib/features/email_previewer/email_previewer_controller.dart

```diff
@@ -0,0 +1,401 @@
+      } else {
+        emlContentViewState.value = Left(failure);
+        _removePreviewEmlContentShared(failure.keyStored!);
+      }
+    } else if (failure is GetPreviewEMLContentInMemoryFailure) {
+      if (failure.exception is NotFoundDataWithThisKeyException) {
+        _keyStored = failure.keyStored;
+        reload();
+      } else {
+        emlContentViewState.value = Left(failure);
+      }
+    } else if (failure is GetCredentialFailure ||
+        failure is GetStoredTokenOidcFailure ||
+        failure is GetAuthenticatedAccountFailure ||
+        failure is GetSessionFailure ||
+        failure is ParseEmailByBlobIdFailure ||
+        failure is PreviewEmailFromEmlFileFailure) {
+      emlContentViewState.value = Left(failure);
+    }else {
+      super.handleFailureViewState(failure);
+    }
+  }
+
+  @override
+  void handleReloaded(Session session) {
+    _session = session;
+    _accountId = session.accountId;
+
+    if (_keyStored != null) {
+      _parseEmailByBlobId(session.accountId, _keyStored!);
+    }
+  }
+
+  void _getPreviewEmailEMLContentShared() {
+    final parameters = Get.parameters;
+
+    if (parameters.isEmpty) {
+      consumeState(Stream.value(Left(GetPreviewEmailEMLContentSharedFailure(
+          exception: NotFoundBlobIdException([])))));
+      return;
+    }
+
+    final keyStored = parameters[RouteUtils.paramID];
+    if (keyStored?.trim().isNotEmpty == true) {
+      consumeState(_getPreviewEmailEMLContentSharedInteractor.execute(keyStored!));
+    } else {
+      consumeState(Stream.value(Left(GetPreviewEmailEMLContentSharedFailure(
+          exception: NotFoundBlobIdException([])))));
+    }
+  }
+
+  void _movePreviewEmlContentFromPersistentToMemory(EMLPreviewer emlPreviewer) {
+    consumeState(
+      _movePreviewEmlContentFromPersistentToMemoryInteractor.execute(emlPreviewer),
+    );
+  }
+
+  void _removePreviewEmlContentShared(String keyStored) {
+    consumeState(_removePreviewEmailEmlContentSharedInteractor.execute(keyStored));
+  }
+
+  void _getPreviewEMLContentInMemory(String keyStored) {
+    consumeState(_getPreviewEmlContentInMemoryInteractor.execute(keyStored));
+  }
+
+  Future<void> onClickHyperLink(Uri? uri) async {
+    log('EmailPreviewerController::onClickHyperLink:Uri = $uri');
+    if (uri == null) return;
+
+    if (uri.scheme == Constant.attachmentScheme) {
+      final attachment = EmailUtils.parsingAttachmentByUri(uri);
+      if (attachment == null) return;
+
+      _downloadAttachment(attachment);
+    } else {
+      _openNewWindowByHyperLink(uri);
+    }
+  }
+
+  void _openNewWindowByHyperLink(Uri uri) {
+    bool isEMlPreview = uri.toString().startsWith(RouteUtils.emailEMLPreviewerRoutePath);
+    final url = _standardizeURL(uri);
+    log('EmailPreviewerController::_openNewWindowByHyperLink: url = $url');
+
+    bool isOpen = HtmlUtils.openNewWindowByUrl(
+      url,
+      isFullScreen: !isEMlPreview,
+      isCenter: false,
+    );
```

Comment:
can we open it in new tab beside the origin tab of tmail?

Reply:
Of course.